docs/source/notes/extending.func.rst

```diff
@@ -0,0 +1,479 @@
+            dx = 3 * x ** 2
+            return result, dx
+
+        @staticmethod
+        def setup_context(ctx, inputs, output):
+            x, = inputs
+            result, dx = output
+            ctx.save_for_backward(x, dx)
+
+        @staticmethod
+        def backward(ctx, grad_output, grad_dx):
+            x, dx = ctx.saved_tensors
+            result = grad_output * dx + grad_dx * 6 * x
+            return result
+
+    def my_cube(x):
+        result, dx = MyCube.apply(x)
+        return result
+
+    x = torch.randn(3)
+    result = torch.vmap(my_cube)(x)
+    assert torch.allclose(result, x ** 3)
+
+
+Defining the vmap staticmethod
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If your :meth:`~torch.autograd.function.FunctionCtx.save_for_backward` calls into another system (like NumPy, C++, CUDA, triton),
+then to get it to work with :func:`torch.vmap` or transforms that use it, you'll
+need to manually define a ``vmap`` staticmethod.
+
+Depending on what transforms you want to use and your use case, you may not need
+to add a ``vmap`` staticmethod to all of your :class:`torch.autograd.Function`:
+
+- For example, :func:`torch.func.jacrev` performs :func:`~torch.vmap` over the backward pass.
+  So if you're only interested in using :func:`torch.func.jacrev`, only
+  the :meth:`~Function.backward` staticmethod needs to be vmappable.
+
+We do recommend adding a ``vmap`` staticmethod to all of your :meth:`~torch.autograd.function.FunctionCtx.save_for_backward`
+though, especially if you are writing a third-party library and you want your
+:meth:`~torch.autograd.function.FunctionCtx.save_for_backward` to work with all combinations of :func:`torch.func` transforms.
+
+Conceptually, the vmap staticmethod is responsible for defining how the :meth:`~Function.forward`
+should behave under :func:`torch.vmap`. That is, it defines how to transform
+the :meth:`~Function.forward` to run over inputs with an additional dimension (the dimension
+being vmapped over). This is similar to how :func:`torch.vmap` is implemented over
+PyTorch operations: for each operation, we define a vmap rule (sometimes also
+referred to as a "batching rule").
+
+Here's how to define the ``vmap`` staticmethod:
+
+- the signature is ``vmap(info, in_dims: Tuple[Optional[int]], *args)``, where
+  ``*args`` is the same as the args to :meth:`~Function.forward`.
+- The vmap staticmethod is responsible for defining how the :meth:`~Function.forward` should behave
+  under :func:`torch.vmap`. That is, given inputs with an additional dimension
+  (specified by ``in_dims``), how do we compute the batched version of :meth:`~Function.forward`?
+- For each arg in ``args``, ``in_dims`` has a corresponding ``Optional[int]``.
+  It is ``None`` if the arg is not a Tensor or if the arg is not being vmapped over,
+  otherwise, it is an integer specifying what dimension of the Tensor is being vmapped
+  over.
+- ``info`` is a collection of additional metadata that may be helpful. For example,
```

Comment:
Do we want to mention all the metadata here? (or link to somewhere where one could find that information)

Reply:
Sure